apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/infer-entities-from-web-page-worker-agent/tool-definitions.ts

```diff
@@ -0,0 +1,394 @@
+          `),
+        },
+        entityTypeIds: {
+          type: "array",
+          items: {
+            type: "string",
+          },
+          description: dedent(`
+            An array of entity type IDs which should be inferred from the provided HTML content.
+          `),
+        },
+        linkEntityTypeIds: {
+          type: "array",
+          items: {
+            type: "string",
+          },
+          description: dedent(`
+            An array of link entity type IDs which should be inferred from provided HTML content.
+          `),
+        },
+      },
+      required: [
+        "url",
+        "hmtlContent",
+        "expectedNumberOfEntities",
+        "validAt",
+        "prompt",
+        "explanation",
+        "entityTypeIds",
+      ],
+    },
+  },
+  submitProposedEntities: {
+    name: "submitProposedEntities",
+    description:
+      "Submit one or more proposed entities as the `result` of the inference task.",
+    inputSchema: {
+      type: "object",
+      properties: {
+        explanation: explanationDefinition,
+        entityIds: {
+          type: "array",
+          items: {
+            type: "string",
+          },
+          description: dedent(`
+            An array of entity IDs of the entities to submit.
+            These must correspond to the IDs provided by a "inferEntitiesFromWebPage" tool call.
+          `),
+        },
+      },
+      required: ["entityIds", "explanation"],
+    },
+  },
+  queryPdf: {
+    name: "queryPdf",
+    description: dedent(`
+      Query a PDF document hosted at a URL.
+      Use this tool to get relevant text out of a PDF document.
+    `),
+    inputSchema: {
+      type: "object",
+      properties: {
+        explanation: explanationDefinition,
+        fileUrl: {
+          type: "string",
+          description: "The absolute URL of the PDF document.",
+        },
+        /**
+         * Note: using the argument name `query` here lead to the LLM agent
+         * providing very short queries, leading to worse results. Using
+         * `description` instead seems to lead to more verbose descriptions
+         * of the information being sought, hence better query results.
+         */
+        description: {
+          type: "string",
+          description: dedent(`
+            A detailed description of the relevant information you are seeking from the PDF document.
+            Include keywords, phrases, or the name specific sections you are looking for.
+            Use language and terminology that is similar to the content in the PDF document you are seeking.
+            If you are looking for specific properties of an entity, name them.
+          `),
+        },
+        exampleText: {
+          type: "string",
+          description: dedent(`
+            An example of the data you are looking for in the PDF document.
+            This can be a table, a paragraph, or any other relevant content.
+          `),
+        },
```

Comment:
Out of interest how does this work in practice? My gut reaction is that it might make up some stuff that bears no resemblance to how the document is written, how tables are formatted etc, or which _does_ but corresponds to the 'wrong' section and thus introduces noise in the matching. But I may be wrong and it helps to find the info

Reply:
You are correct that the agent makes up things here. It's usually a lit of information that it's looking for a given entity, i.e. something like "Investment Fund X, Y voting rights, Z% ownership," etc).

This was the only way I managed to get the worker agent to compose a query that actually returns the relevant table via the similarity search within 15 chunks, most of the time. My understanding of why this works better than another generic query is because the naming the properties in the query results in a closer embedding between the query and the relevant chunk.

I've attempted to ask the agent to specify the properties it's looking for in the query, but it often times didn't make such an attempt.

If you have other ideas on alternative approaches we could use to getting the worker agent to produce a better query, let me know.

Reply:
Not currently, no! Fair enough